### ISSUE TITLE
Op/optimization TODO: log determinant

### ISSUE THREAD

**Issue by dwf** [Theano/Theano#150]
Now that we have a determinant Op we should probably look into this.

There are numerically stable ways to compute the natural logarithm of the determinant, a quantity which is often needed in evaluating the log probability of a multivariate Gaussian distribution. NumPy provides this as [`numpy.slogdet`](http://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.slogdet.html) for "sign log determinant" (it computes `log(abs(det(X))`) and also returns the sign of the determinant).

The derivative of the log determinant also has a particularly simple form as `inv(X).T == inv(X.T)`.

**Comment by harpone** [comment]
Just out of curiosity, why hasn't this been done yet? I haven't coded any Theano ops, but I could give it a shot... maybe just a log abs det, since that's probably the most common case (in MLE estimation etc)?

FYI I'm hitting the stability issue pretty badly...

**Comment by lamblin** [comment]
I'm not sure why this has not been done yet, or maybe some people ended up implementing that in their own repositories.
If you want to give it a shot, it would be welcome. There should be a section on how to extend theano in the online doc. As long as you use python/numpy for the implementation and do not need a GPU version, it should be quite straightforward.
Please let us know if you have any issues or questions, either in a pull request or at `theano-dev`.

**Comment by harpone** [comment]
OK I'll give it a try!

**Comment by harpone** [comment]
Found a couple of implementations on Github. [This one](https://github.com/FlorianSeidel/GOL/blob/9f7716e78a78fbdd7ae8ecac5a58b701f6a4da41/gol/goal/constraints.py) does an SVD first and takes the log of the diagonal (not sure why it's diagonal squared... need to check the math; also has a funny expression for the gradient... it's correct but computationally expensive).

[This one](https://github.com/kolia/subunits/blob/05f40cbc45214d80567b132c95f16913ca2a0cc7/kolia_theano.py#L12) uses numpy's slogdet.

Any preferences? SVD on GPU should be a simple copy/paste job I think, but I haven't found any CUDA/PyCUDA/gnumpy code for the slogdet...

**Comment by nouiz** [comment]
You could check inside numpy code how slogdet is implemented. Maybe it does
it with an svd? Can you compate the output of both op to see if the result
is the same?

There is svd code on the GPU somewhere, so with that, it wouldn't be too
hard to make a GPU implementation

**Comment by harpone** [comment]
It calls something in lapack routines in numpy/numpy/linalg/umath_linalg.c.src, which says in docstring that it "computes logdet from factored diagonal" so I guess it is SVD (or really an eigendecomposition since it's for square matrices)...

The square in the svd code was wrong as I suspected, but fixing that, the SVD and slogdet versions give slightly different results! Difference of about 1E-5 for 100*100 random normal matrices about 80% of time and zero difference 20% of time... happens for pure numpy/scipy versions too. Maybe there's some numerical stability issues in first doing the svd, then taking log and then sum??

I'll post the code below if anyone wants to take a look. I'm actually taking log abs det, because why take a log of a negative number...

So I'm guessing the slogdet version is numerically more accurate, which raises the question of how to do it accurately on a GPU... dammit!

PS. This will (hopefully) be my first ever PR, so I need some time to figure out how to code the tests, where to put them etc...

`#NUMPY SLOGDET:
from theano.gof import Op, Apply
import numpy

matrix_inverse = T.nlinalg.MatrixInverse()

class LogAbsDet(Op):
    """Logarithm of absolute determinant of a matrix M,
    log(abs(det(M))). Avoids det(M) overflow/ underflow.
    """
    def make_node(self, x):
        x = theano.tensor.as_tensor_variable(x)
        o = theano.tensor.scalar(dtype=x.dtype)
        return Apply(self, [x], [o])
    
    def perform(self, node, (x,), (z,)):
        try:
            _, logabsdet = numpy.linalg.slogdet(x)
            z[0] = numpy.asarray(logabsdet, dtype=x.dtype)
        except Exception:
            print('Failed to compute determinant of {}.'.format(x))
            raise
            
    def grad(self, inputs, g_outputs):
        gz, = g_outputs
        x, = inputs
        return [gz * matrix_inverse(x).T]
    
    def __str__(self):
        return "LogAbsDet"
    
logabsdet = LogAbsDet()`

The SVD is the same, except now there's this:

`    s = svd(x, compute_uv=False)
     z[0] = np.asarray(np.sum(np.log(s)), dtype=x.dtype)`

**Comment by harpone** [comment]
For float64 the error is 1E-14:

`from numpy.linalg import slogdet
from numpy.linalg import svd

testmat = np.random.randn(n, n)

_, logabsdet_numpy_test_slogdet = slogdet(testmat)
logabsdet_numpy_test_svd = np.sum(np.log(svd(testmat, compute_uv=False)))

logabsdet_numpy_test_slogdet - logabsdet_numpy_test_svd`

For float32 the error is 1E-5!!

`from numpy.linalg import slogdet
from numpy.linalg import svd

testmat = np.random.randn(n, n).astype(np.float32)

_, logabsdet_numpy_test_slogdet = slogdet(testmat)
logabsdet_numpy_test_svd = np.sum(np.log(svd(testmat, compute_uv=False)))

logabsdet_numpy_test_slogdet - logabsdet_numpy_test_svd`

Anyone know the reason for this?

**Comment by harpone** [comment]
Given that the logdet _values_ for the examples above are of order 1E+2, the relative error is around 1E-7...1E-8 for float32... maybe that's acceptable?

I think I should go with the direct SVD method then for the 'perform' method, if the GPU version is going to be also SVD...

**Comment by lamblin** [comment]
Yes, this kind of differences is really in the normal range for float32, these are even quite low.

**Comment by harpone** [comment]
OK did a PR here: https://github.com/Theano/Theano/pull/3954

By the way, SVD on GPU seems to be useful only when matrix dims go above 1000 * 1000, so maybe the GPU implementation is not that important yet? Or will there be some overhead from loading parameters back and forth between GPU memory and RAM??

This image is from [this paper](http://www.pascal-man.com/navigation/faq-java-browser/GPU/Sheetal09Singular.pdf):

![image](https://cloud.githubusercontent.com/assets/5112840/12696295/4f59bd1c-c76f-11e5-9346-43f9dfc54ded.png)

**Comment by harpone** [comment]
New PR due to me being a n00b: https://github.com/Theano/Theano/pull/3959

**Comment by nouiz** [comment]
The transfer to/from the GPU/CPU is to be avoided as much as possible. So
even if the op isn't faster, just for that reason, it can be useful to add
it.

**Comment by harpone** [comment]
OK got it
